README.md
# strong-soap

[![CI](https://github.com/loopbackio/strong-soap/actions/workflows/continuous-integration.yaml/badge.svg)](https://github.com/loopbackio/strong-soap/actions/workflows/continuous-integration.yaml)
[![Build Status](https://app.travis-ci.com/loopbackio/strong-soap.svg?branch=master)](https://app.travis-ci.com/loopbackio/strong-soap)
[![Coverage Status](https://coveralls.io/repos/github/loopbackio/strong-soap/badge.svg?branch=master)](https://coveralls.io/github/loopbackio/strong-soap?branch=master)

This module provides a Node.js SOAP client for invoking web services and a mock-up SOAP server capability to create and test your web service. This module is based on `node-soap` module.

<!-- Run `npm run toc` to update below section -->
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [strong-soap](#strong-soap)
  - [Overview](#overview)
  - [Install](#install)
  - [Client](#client)
    - [Extra headers (optional)](#extra-headers-optional)
    - [Client.describe()](#clientdescribe)
    - [Client.setSecurity(security)](#clientsetsecuritysecurity)
    - [Client._method_(args, callback)](#clientmethodargs-callback)
    - [Client._service_._port_._method_(args, callback[, options[, extraHeaders]])](#clientserviceportmethodargs-callback-options-extraheaders)
      - [Options (optional)](#options-optional)
      - [Alternative method call using callback-last pattern](#alternative-method-call-using-callback-last-pattern)
    - [Client._lastRequest_](#clientlastrequest)
    - [Client.setEndpoint(url)](#clientsetendpointurl)
    - [Client events](#client-events)
    - [Security](#security)
    - [BasicAuthSecurity](#basicauthsecurity)
    - [BearerSecurity](#bearersecurity)
    - [ClientSSLSecurity](#clientsslsecurity)
    - [WSSecurity](#wssecurity)
    - [WSSecurityCert](#wssecuritycert)
    - [ClientSSLSecurityPFX](#clientsslsecuritypfx)
  - [XML attributes](#xml-attributes)
    - [Handling XML attributes, value, and XML (wsdlOptions)](#handling-xml-attributes-value-and-xml-wsdloptions)
    - [Overriding the value key](#overriding-the-value-key)
    - [Overriding the xml key](#overriding-the-xml-key)
    - [Overriding the attributes key](#overriding-the-attributes-key)
  - [XMLHandler](#xmlhandler)
  - [WSDL](#wsdl)
    - [wsdl.open(wsdlURL, options, callback(err, wsdl))](#wsdlopenwsdlurl-options-callbackerr-wsdl)
    - [wsdl.openSync(wsdlURL, options)](#wsdlopensyncwsdlurl-options)
  - [Server](#server)
    - [soap.listen(_server_, _path_, _services_, _wsdl_)](#soaplistenserver-path-services-wsdl)
    - [Options](#options)
    - [Server logging](#server-logging)
    - [Server events](#server-events)
    - [SOAP Fault](#soap-fault)
    - [Server security example using PasswordDigest](#server-security-example-using-passworddigest)
    - [Server connection authorization](#server-connection-authorization)
  - [SOAP headers](#soap-headers)
    - [Received SOAP headers](#received-soap-headers)
    - [Outgoing SOAP headers](#outgoing-soap-headers)
      - [addSoapHeader(value, qname)](#addsoapheadervalue-qname)
      - [changeSoapHeader(index, value, qname)](#changesoapheaderindex-value-qname)
      - [getSoapHeaders()](#getsoapheaders)
      - [clearSoapHeaders()](#clearsoapheaders)
  - [soap-stub](#soap-stub)
    - [Example](#example)
  - [Contributors](#contributors)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

Features:

- Full SOAP Client capability and mock-up SOAP server capability
- Handles both RPC and Document styles
- Handles both SOAP 1.1 and SOAP 1.2 Fault
- APIs to parse XML into JSON and JSON into XML
- API to describe WSDL document
- Support for both synchronous and asynchronous method handlers
- WS-Security (currently only UsernameToken and PasswordText encoding is supported)

## Install

Node.js version 10, 12, and 14 are officially supported. We dropped version 8
support in 3.0.0.

Install with [npm](http://github.com/isaacs/npm):

```sh
npm install strong-soap
```

## Client

Start with the WSDL for the web service you want to invoke. For example, the stock quote service http://www.webservicex.net/stockquote.asmx and the WSDL is http://www.webservicex.net/stockquote.asmx?WSDL

Create a new SOAP client from WSDL URL using `soap.createClient(url[, options], callback)`. Also supports a local file system path. An instance of `Client` is passed to the `soap.createClient` callback. It is used to execute methods on the soap service.

```js
"use strict";

var soap = require("strong-soap").soap;
// wsdl of the web service this client is going to invoke. For local wsdl you can use, url = './wsdls/stockquote.wsdl'
var url = "http://www.webservicex.net/stockquote.asmx?WSDL";

var requestArgs = {
  symbol: "IBM",
};

var options = {};
soap.createClient(url, options, function (err, client) {
  var method = client["StockQuote"]["StockQuoteSoap"]["GetQuote"];
  method(requestArgs, function (err, result, envelope, soapHeader) {
    //response envelope
    console.log("Response Envelope: \n" + envelope);
    //'result' is the response body
    console.log("Result: \n" + JSON.stringify(result));
  });
});
```

As well as creating a client via a `url`, an existing [WSDL](#wsdl) object can be passed in via `options.WSDL_CACHE`.

```js
var soap = require("strong-soap").soap;
var WSDL = soap.WSDL;

var url = "http://www.webservicex.net/stockquote.asmx?WSDL";

// Pass in WSDL options if any

var options = {};
WSDL.open(url, options, function (err, wsdl) {
  // You should be able to get to any information of this WSDL from this object. Traverse
  // the WSDL tree to get  bindings, operations, services, portTypes, messages,
  // parts, and XSD elements/Attributes.

  // Set the wsdl object in the cache. The key (e.g. 'stockquotewsdl')
  // can be anything, but needs to match the parameter passed into soap.createClient()
  var clientOptions = {
    WSDL_CACHE: {
      stockquotewsdl: wsdl,
    },
  };
  soap.createClient("stockquotewsdl", clientOptions, function (err, client) {
    var method = client["StockQuote"]["StockQuoteSoap"]["GetQuote"];
    method(requestArgs, function (err, result, envelope, soapHeader) {
      //response envelope
      console.log("Response Envelope: \n" + envelope);
      //'result' is the response body
      console.log("Result: \n" + JSON.stringify(result));
    });
  });
});
```

The Request envelope created by above service invocation:

```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
  <soap:Header/>
  <soap:Body>
    <ns1:GetQuote xmlns:ns1="http://www.webserviceX.NET/">
      <ns1:symbol>IBM</ns1:symbol>
    </ns1:GetQuote>
  </soap:Body>
</soap:Envelope>
```

This WSDL operation is defined as document/literal-wrapped style. Hence the request in soap <Body> is wrapped in operation name. Refer to test cases [server-client-document-test](https://github.com/loopbackio/strong-soap/blob/master/test/server-client-document-test.js) and [server-client-rpc-test](https://github.com/loopbackio/strong-soap/blob/master/test/server-client-rpc-test.js) to understand document and rpc styles and their
Request, Response and Fault samples.

The `options` argument allows you to customize the client with the following properties:

- `endpoint`: to override the SOAP service's host specified in the `.wsdl` file.
- `request`: to override the [request](https://github.com/request/request) module.
- `httpClient`: to provide your own http client that implements `request(rurl, data, callback, exheaders, exoptions)`.
- `envelopeKey`: to set specific key instead of <pre><<b>soap</b>:Body></<b>soap</b>:Body></pre>
- `wsdl_options`: custom options for the request module on WSDL requests.
- `wsdl_headers`: custom HTTP headers to be sent on WSDL requests.

Note: for versions of node >0.10.X, you may need to specify `{connection: 'keep-alive'}` in SOAP headers to avoid truncation of longer chunked responses.

### Extra headers (optional)

User can define extra HTTP headers to be sent on the request.

```js
var clientOptions = {};
soap.createClient(url, clientOptions, function (err, client) {
  var customRequestHeader = { customheader1: "test1" };
  // Custom request header
  client.GetQuote(
    requestArgs,
    function (err, result, envelope) {
      // Result in SOAP envelope body which is the wrapper element.
      // In this case, result object corresponds to GetCityForecastByZIPResponse.
      console.log(JSON.stringify(result));
    },
    null,
    customRequestHeader
  );
});
```

### Client.describe()

Describes services, ports and methods as a JavaScript object.

```js
// Describes the entire WSDL in a JSON tree object form.
var description = client.describe();
// Inspect GetQuote operation. You can inspect Service: {Port: {operation: {
console.log(JSON.stringify(description.StockQuote.StockQuoteSoap.GetQuote));
```

### Client.setSecurity(security)

Use the specified security protocol.

Refer to test case [ssl-test](https://github.com/loopbackio/strong-soap/blob/master/test/ssl-test.js) for an example of using this API.

### Client._method_(args, callback)

Call _method_ on the SOAP service.

```js
client.MyFunction(
  { name: "value" },
  function (err, result, envelope, soapHeader) {
    // Result is a javascript object
    // Envelope is the response envelope from the Web Service
    // soapHeader is the response soap header as a JavaScript object
  }
);
```

A _method_ can also be called as a promise.

```js
client.MyFunction({ name: "value" }).then(
  function ({ result, envelope, soapHeader }) {
    // ...
  },
  function (err) {
    // ...
  }
);

// in async/await flavor
try {
  const { result, envelope, soapHeader } = await client.MyFunction({
    name: "value",
  });
} catch (err) {
  // handle error
}
```

### Client._service_._port_._method_(args, callback[, options[, extraHeaders]])

Call a _method_ using a specific _service_ and _port_.

```js
client.MyService.MyPort.MyFunction({ name: "value" }, function (err, result) {
  // Result is a JavaScript object
});
```

#### Options (optional)

Accepts any option that the request module accepts, see [request](https://github.com/request/request) module.

For example, you could set a timeout of 5 seconds on the request like this:

```js
client.MyService.MyPort.MyFunction(
  { name: "value" },
  function (err, result) {
    // result is a javascript object
  },
  { timeout: 5000 }
);
```

You can measure the elapsed time on the request by passing the time option:

```js
client.MyService.MyPort.MyFunction(
  { name: "value" },
  function (err, result) {
    // client.lastElapsedTime - the elapsed time of the last request in milliseconds
  },
  { time: true }
);
```

#### Alternative method call using callback-last pattern

To align method call signature with Node's standard callback-last pattern and eventually allow promisification of method calls, the following method signatures are also supported:

```js
client.MyService.MyPort.MyFunction(
  { name: "value" },
  options,
  function (err, result) {
    // result is a javascript object
  }
);

client.MyService.MyPort.MyFunction(
  { name: "value" },
  options,
  extraHeaders,
  function (err, result) {
    // result is a javascript object
  }
);
```

### Client._lastRequest_

The property that contains last full soap request for client logging.

### Client.setEndpoint(url)

Overwrites the SOAP service endpoint address.

### Client events

Client instances emit the following events:

- request - Emitted before a request is sent. The event handler receives the
  entire Soap request (Envelope) including headers.
- message - Emitted before a request is sent. The event handler receives the
  Soap body contents. Useful if you don't want to log /store Soap headers.
- soapError - Emitted when an erroneous response is received.
  Useful if you want to globally log errors.
- response - Emitted after a response is received. The event handler receives
  the SOAP response body as well as the entire `IncomingMessage` response object.
  This is emitted for all responses (both success and errors).

For an example of using this API, see [ssl-test](https://github.com/loopbackio/strong-soap/blob/master/test/client-test.js).

Here is an example of 'soapError' event

```js
soap.createClient(
  __dirname + "/wsdl/default_namespace.wsdl",
  function (err, client) {
    var didEmitEvent = false;
    client.on("soapError", function (err) {
      didEmitEvent = true;
      assert.ok(err.root.Envelope.Body.Fault);
    });
    client.MyOperation({}, function (err, result) {
      assert.ok(didEmitEvent);
      done();
    });
  },
  baseUrl
);
```

### Security

`strong-soap` has several default security protocols. You can easily add your own
as well. The interface is quite simple. Each protocol defines two methods:

- `addOptions` - Method that accepts an options arg that is eventually passed directly to `request`
- `toXML` - Method that returns a string of XML.

### BasicAuthSecurity

```js
client.setSecurity(new soap.BasicAuthSecurity("username", "password"));
```

### BearerSecurity

```js
client.setSecurity(new soap.BearerSecurity("token"));
```

### ClientSSLSecurity

_Note_: If you run into issues using this protocol, consider passing these options
as default request options to the constructor:

- `rejectUnauthorized: false`
- `strictSSL: false`
- `secureOptions: constants.SSL_OP_NO_TLSv1_2` (this is likely needed for node >= 10.0)

```js
client.setSecurity(
  new soap.ClientSSLSecurity("/path/to/key", "/path/to/cert", {
    /*default request options*/
  })
);
```

### WSSecurity

`WSSecurity` implements WS-Security. UsernameToken and PasswordText/PasswordDigest is supported.

```js
var wsSecurity = new WSSecurity(username, password, options);
//the 'options' object is optional and contains properties:
//passwordType: 'PasswordDigest' or 'PasswordText' default is PasswordText
//hasTimeStamp: true or false, default is true
//hasTokenCreated: true or false, default is true
client.setSecurity(wsSecurity);
```

### WSSecurityCert

WS-Security X509 Certificate support.

```js
var privateKey = fs.readFileSync(privateKeyPath);
var publicKey = fs.readFileSync(publicKeyPath);
var password = ""; // optional password
var wsSecurity = new soap.WSSecurityCert(
  privateKey,
  publicKey,
  password,
  "utf8"
);
client.setSecurity(wsSecurity);
```

_Note_: Optional dependency 'ursa' is required to be installed successfully when WSSecurityCert is used.

### ClientSSLSecurityPFX

```js
const pfxSecurity = new soap.ClientSSLSecurityPFX(
  pathToPfxOrFileBuffer,
  passphrase
);
client.setSecurity(pfxSecurity);
```

## XML attributes

### Handling XML attributes, value, and XML (wsdlOptions)

To override the default behavior of `strong-soap`, use the `wsdlOptions` object, passed in the
`createClient()` method. The `wsdlOptions` has the following properties:

```js
var wsdlOptions = {
  attributesKey: "theAttrs",
  valueKey: "theVal",
  xmlKey: "theXml",
};
```

If you call `createClient()` with no options (or an empty Object `{}`), `strong-soap` defaults
to the following:

- `attributesKey` : `'$attributes'`
- `valueKey` : `'$value'`
- `xmlKey` : `'$xml'`

### Overriding the value key

By default, `strong-soap` uses `$value` as key for any parsed XML value which may interfere with your other code as it
could be some reserved word, or the `$` in general cannot be used for a key to start with.


CHANGES.md
# 2021-09-13, Version 3.3.1

- fix: fixed installing the library from npm (Aliaksandr Krasitski)

# 2021-09-13, Version 3.3.0

- Typo in example server code. (Sam Graham)

- chore: enable sourcemaps (Rifa Achrinza)

- docs: add `ClientSSLSecurityPFX` to readme (Luan Raithz Machado)

- chore: bump xml-crypto 2.1.3 (Aliaksandr Krasitski)

# 2021-04-22, Version 3.2.1

- Update pakcage lock (Jonathan Roberts)

- NTLM fixes and new tests (Jonathan Roberts)

# 2021-04-07, Version 3.2.0

# 2021-04-07, Version 3.1.0

- chore: update dependencies (Diana Lau)

- add require globalize to fix broken call to g.f() (Stefan Seide)

# 2021-03-29, Version 3.0.1

- bump/xml crypto 2.1.1 (Daniel Hillmann)

# 2020-10-28, Version 3.0.0

- feat: update dependencies (Raymond Feng)

- chore: update lodash (jannyHou)

- chore: update CODEOWNERS (Diana Lau)

- feat(client) add soapHeaders at run time (Mario Estrada)

- Handle restrictions in anonymous simple types (Jonathan Roberts)

# 2020-03-25, Version 2.0.0

- Update client.js: add RPC suffixes (kimveasna)

- In light of review, check isMany in existing block (LoneRifle)

- Add tests for responses when maxOccurs=1 (LoneRifle)

- xmlToJson: honor maxOccurs > 1 (LoneRifle)

- De-lint - missing semicolons (LoneRifle)

# 2020-02-25, Version 1.22.1

- Allow nillable date to be null (Richard Sunbury)

- chore: remove unused require statements (Raymond Feng)

# 2020-02-10, Version 1.22.0

- chore: add package-lock.json (Raymond Feng)

- Remove unnecessary files from npm packages (Avindra Goolcharan)

- fixed formating in README (Christopher Wade)

- chore: add stalebot (Diana Lau)

- Avoid line feed problems (Win/Linux) (Christian Herzog)

# 2019-07-19, Version 1.21.0

- add support attachments (Allan Marques)

# 2019-06-03, Version 1.20.0

- Replace ursa with node 12 core crypto or node-rsa (Raymond Feng)

- chore: update copyrights years (Agnes Lin)

- Add support for 1/0 booleans (Stefano Marotta)

- fix: update lodash (jannyHou)

# 2019-05-02, Version 1.19.1

- Use Buffer.from to replace depreated new Buffer (Raymond Feng)

- Fix xml date/time/dateTime conversion (Raymond Feng)

- Replace deprecated Buffer api (Raymond Feng)

- Add promise based stubs for testing (Rijnhard Hessel)

# 2019-04-29, Version 1.19.0

- Upgrade dependencies (Raymond Feng)

# 2019-03-28, Version 1.18.0

- chore: upgrade deps (Raymond Feng)

- Tests for parseValue with xsd date dateTime (arthmoeros)

- Fix for xs:date with tz format (arthmoeros)

# 2019-02-21, Version 1.17.0

- Fix use options.request in wsdl.js (JB)

# 2019-02-14, Version 1.16.0

- Update version of httpntlm (Jon Roberts)

# 2019-01-22, Version 1.15.0

- Adding enforce restrictions option (Romeu Palos de Gouvea)

# 2018-10-18, Version 1.14.0

- Update CODEOWNERS (Raymond Feng)

- Rename test so it gets run (Jon Roberts)

- Return error via callback (Jon Roberts)

# 2018-10-12, Version 1.13.2

- changed this.soapActionRequired to boolean (Vineet Jain)

- Support of sending soapAction in v1.2 (Vineet Jain)

# 2018-09-24, Version 1.13.1

- Add tests to cover maxOccurs (Jon Roberts)

- Change to isMany logic (Jon Roberts)

# 2018-09-17, Version 1.13.0

- Update ssl key/cert to pass node 10 (Raymond Feng)

- Update` strong-globalize` to v4.1.1 (Runrioter Wung)

- Switch to strong-ursa to support Node 10 (Raymond Feng)

- Add ability to load in a sync way (DomStorey)

- Add the ability to loadSync (DomStorey)

- fix code blocks in readme.md with their type (Jiri Spac)

- fix: use proper variable names (biniam)

# 2018-08-08, Version 1.12.1

- Allow documentation tag under types (Jon Roberts)

# 2018-08-06, Version 1.12.0

- Create client from cached wsdl (Jon Roberts)

- add forgotten options to soap.listen(...) (Vladimir Latyshev)

# 2018-07-10, Version 1.11.0

- Load a wsdl with no imports from the cache (Jon Roberts)

# 2018-07-09, Version 1.10.0

- [WebFM] cs/pl/ru translation (candytangnb)

- fix soap-stub example on README.md (Carlos Assis)

# 2018-05-08, Version 1.9.0

# 2018-05-07, Version 1.8.0

- feat: add promise support for client operation (shimks)

# 2018-04-12, Version 1.7.0

- fix: add element text value (Raymond Feng)

# 2018-03-26, Version 1.6.2

- fix xsd include/import recursive processing (Raymond Feng)

# 2018-03-19, Version 1.6.1

- fix: add array check for concat (Raymond Feng)

- build: drop node 4.x support (Raymond Feng)

# 2018-03-16, Version 1.6.0

- fix: improve wsdl types processing for multiple schemas (Raymond Feng)

- chore: clean up deps (Raymond Feng)

# 2018-01-31, Version 1.5.0

- fix: upgrade deps (Raymond Feng)

- fix: make sure xml elements are mapped based the order of xsd (Raymond Feng)

- Fix Typo: RPC Literal (Thomas Richter)

- check the original passwordDigest method fails against SoapUI output (Tom Hodder)

- revert formatting changes by format on save in vscode (Tom Hodder)

- Updated passwordDigest method and related tests (Tom Hodder)

- missing semi colons (Branden Horiuchi)

- removed lodash dependency (Branden Horiuchi)

- added CookieSecurity test (Branden Horiuchi)

- better cookie parsing handle the case where the header for set-cookie is a string and not an array (Branden Horiuchi)

- fixing header parse objects would not have been parsed (Branden Horiuchi)

- added cookie security (Branden Horiuchi)

# 2018-01-24, Version 1.4.2

- fix: handle xsi:type for subtype mappng from json to xml (Raymond Feng)

# 2018-01-15, Version 1.4.1

- make sure attribute descriptors are honored to map json to xml (Raymond Feng)

# 2018-01-12, Version 1.4.0

- allow xsd:include to include a schema without target namespace (Raymond Feng)

# 2018-01-03, Version 1.3.0

- remove node 0.10/0.12 support (Raymond Feng)

- fix soap version detection (Raymond Feng)

- Create Issue and PR Templates (#124) (Sakib Hasan)

- Add CODEOWNER file (Diana Lau)

- update translation file (Diana Lau)

# 2017-06-28, Version 1.2.6

- removed process.env.PORT (rashmihunt)

- fix ports (rashmihunt)


src/server.js
// Copyright IBM Corp. 2016,2018. All Rights Reserved.
// Node module: strong-soap
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

// @ts-check
"use strict";

var g = require("./globalize");
var url = require("url"),
  compress = null,
  events = require("events"),
  XMLHandler = require("./parser/xmlHandler"),
  Base = require("./base"),
  toXMLDate = require("./utils").toXMLDate,
  util = require("util"),
  debug = require("debug")("strong-soap:server"),
  debugDetail = require("debug")("strong-soap:server:detail");

try {
  compress = require("compress");
} catch (error) {
  // Ignore error
}

class Server extends Base {
  constructor(server, path, services, wsdl, options) {
    super(wsdl, options);
    var self = this;
    options = options || {};
    this.path = path;
    this.services = services;

    debug(
      "Server parameters: path: %s services: %j wsdl: %j",
      path,
      services,
      wsdl
    );
    if (path[path.length - 1] !== "/") path += "/";
    wsdl.load(function (err) {
      if (err) throw err;
      self.xmlHandler = new XMLHandler(
        self.wsdl.definitions.schemas,
        self.wsdl.options
      );
      var listeners = server.listeners("request").slice();

      server.removeAllListeners("request");
      server.addListener("request", function (req, res) {
        if (typeof self.authorizeConnection === "function") {
          if (!self.authorizeConnection(req.connection.remoteAddress)) {
            res.end();
            return;
          }
        }
        var reqPath = url.parse(req.url).pathname;
        if (reqPath[reqPath.length - 1] !== "/") reqPath += "/";
        if (path === reqPath) {
          self._requestListener(req, res);
        } else {
          for (var i = 0, len = listeners.length; i < len; i++) {
            listeners[i].call(this, req, res);
          }
        }
      });
    });
  }

  _requestListener(req, res) {
    var self = this;
    var reqParse = url.parse(req.url);
    var reqPath = reqParse.pathname;
    var reqQuery = reqParse.search;

    if (typeof self.log === "function") {
      self.log("info", "Handling " + req.method + " on " + req.url);
    }

    if (req.method === "GET") {
      if (reqQuery && reqQuery.toLowerCase() === "?wsdl") {
        if (typeof self.log === "function") {
          self.log("info", "Wants the WSDL");
        }
        res.setHeader("Content-Type", "application/xml");
        res.write(self.wsdl.toXML());
      }
      res.end();
    } else if (req.method === "POST") {
      res.setHeader("Content-Type", req.headers["content-type"]);
      var chunks = [],
        gunzip;
      if (compress && req.headers["content-encoding"] === "gzip") {
        gunzip = new compress.Gunzip();
        gunzip.init();
      }
      req.on("data", function (chunk) {
        if (gunzip) chunk = gunzip.inflate(chunk, "binary");
        chunks.push(chunk);
      });
      req.on("end", function () {
        var xml = chunks.join("");
        var result;
        var error;
        if (gunzip) {
          gunzip.end();
          gunzip = null;
        }
        try {
          if (typeof self.log === "function") {
            self.log("received", xml);
          }
          self._process(xml, req, function (result, statusCode) {
            if (statusCode) {
              res.statusCode = statusCode;
            }
            res.write(result);
            res.end();
            if (typeof self.log === "function") {
              self.log("replied", result);
            }
          });
        } catch (err) {
          error = err.stack || err;
          res.statusCode = 500;
          res.write(error);
          res.end();
          if (typeof self.log === "function") {
            self.log("error", error);
          }
        }
      });
    } else {
      res.end();
    }
  }

  _process(input, req, callback) {
    var self = this,
      pathname = url.parse(req.url).pathname.replace(/\/$/, ""),
      obj = this.xmlHandler.xmlToJson(null, input),
      body = obj.Body,
      headers = obj.Header,
      bindings = this.wsdl.definitions.bindings,
      binding,
      operation,
      operationName,
      serviceName,
      portName,
      includeTimestamp =
        obj.Header && obj.Header.Security && obj.Header.Security.Timestamp;

    if (typeof self.authenticate === "function") {
      if (!obj.Header || !obj.Header.Security) {
        throw new Error(g.f("No security header"));
      }
      if (!self.authenticate(obj.Header.Security)) {
        throw new Error(g.f("Invalid username or password"));
      }
    }

    if (typeof self.log === "function") {
      self.log("info", "Attempting to bind to " + pathname);
    }

    // use port.location and current url to find the right binding
    binding = (function (self) {
      var services = self.wsdl.definitions.services;
      var firstPort;
      var name;
      for (name in services) {
        serviceName = name;
        var service = services[serviceName];
        var ports = service.ports;
        for (name in ports) {
          portName = name;
          var port = ports[portName];
          var portPathname = url
            .parse(port.location)
            .pathname.replace(/\/$/, "");

          if (typeof self.log === "function") {
            self.log(
              "info",
              "Trying " + portName + " from path " + portPathname
            );
          }

          if (portPathname === pathname) return port.binding;

          // The port path is almost always wrong for generated WSDLs
          if (!firstPort) {
            firstPort = port;
          }
        }
      }
      return !firstPort ? void 0 : firstPort.binding;
    })(this);

    if (!binding) {
      throw new Error(g.f("Failed to bind to {{WSDL}}"));
    }

    try {
      if (binding.style === "rpc") {
        operationName = Object.keys(body)[0];

        self.emit("request", obj, operationName);
        if (headers) self.emit("headers", headers, operationName);

        self._executeMethod(
          {
            serviceName: serviceName,
            portName: portName,
            operationName: operationName,
            outputName: operationName + "Response",
            args: body[operationName],
            headers: headers,
            style: "rpc",
          },
          req,
          callback
        );
      } else {
        //document style
        var messageElemName =
          Object.keys(body)[0] === "attributes"
            ? Object.keys(body)[1]
            : Object.keys(body)[0];
        var pair = binding.topElements[messageElemName];

        var operationName, outputName;
        var operations = binding.operations;
        //figure out the output name
        for (var name in operations) {
          var inputParts = operations[name].input.message.parts;
          //find the first part of the input message. There could be more than one parts in input message.
          var firstInPart = inputParts[Object.keys(inputParts)[0]];
          if (firstInPart.element.$name === messageElemName) {
            operationName = operations[name].$name;
            if (operations[name].output != null) {
              var outPart = operations[name].output.message.parts;
              //there will be only one output part
              var firstOutPart = outPart[Object.keys(outPart)[0]];
              outputName = firstOutPart.element.$name;
            }
            break;
          }
        }

        self.emit("request", obj, operationName);
        if (headers) self.emit("headers", headers, operationName);

        self._executeMethod(
          {
            serviceName: serviceName,
            portName: portName,
            operationName: operationName,
            outputName: outputName,
            args: body[messageElemName],
            headers: headers,
            style: "document",
            includeTimestamp: includeTimestamp,
          },
          req,
          callback
        );
      }
    } catch (error) {
      if (error.Fault !== undefined) {
        return self._sendError(
          operations[name],
          error,
          callback,
          includeTimestamp
        );
      }
      //Revisit - is this needed?
      throw error;
    }
  }

  _executeMethod(options, req, callback) {
    options = options || {};
    var self = this,
      operation,
      body,
      serviceName = options.serviceName,
      portName = options.portName,
      operationName = options.operationName,
      outputName = options.outputName,
      args = options.args,
      style = options.style,
      includeTimestamp = options.includeTimestamp,
      handled = false;

    try {
      operation = this.services[serviceName][portName][operationName];
      debug("Server operation: %s ", operationName);
    } catch (error) {
      debug("Server executeMethod: error: %s ", error.message);
      //fix - should create a fault and call sendError (..) so that this error is not lost and will be sent as Fault in soap envelope
      //to the client?
      return callback(this._envelope("", includeTimestamp));
    }

    function handleResult(error, result) {
      if (handled) return;
      handled = true;

      var operation =
        self.wsdl.definitions.services[serviceName].ports[portName].binding
          .operations[operationName];

      if (error && error.Fault !== undefined) {
        return self._sendError(operation, error, callback, includeTimestamp);
      } else if (result === undefined) {
        // Backward compatibility to support one argument callback style
        result = error;
      }

      var element = operation.output;

      var operationDescriptor = operation.describe(self.wsdl.definitions);
      debugDetail(
        "Server handleResult. operationDescriptor: %j ",
        operationDescriptor
      );

      var outputBodyDescriptor = operationDescriptor.output.body;
      debugDetail(
        "Server handleResult. outputBodyDescriptor: %j ",
        outputBodyDescriptor
      );

      var soapNsURI = "http://schemas.xmlsoap.org/soap/envelope/";
      var soapNsPrefix = self.wsdl.options.envelopeKey || "soap";

      if (operation.soapVersion === "1.2") {
        soapNsURI = "http://www.w3.org/2003/05/soap-envelope";
      }

      debug("Server soapNsURI: %s soapNsPrefix: %s", soapNsURI, soapNsPrefix);

      var nsContext = self.createNamespaceContext(soapNsPrefix, soapNsURI);
      var envelope = XMLHandler.createSOAPEnvelope(soapNsPrefix, soapNsURI);

      self.xmlHandler.jsonToXml(
        envelope.body,
        nsContext,
        outputBodyDescriptor,
        result
      );

      self._envelope(envelope, includeTimestamp);
      var message = envelope.body.toString({ pretty: true });
      var xml = envelope.doc.end({ pretty: true });

      debug("Server handleResult. xml: %s ", xml);
      callback(xml);
    }

    if (
      !self.wsdl.definitions.services[serviceName].ports[portName].binding
        .operations[operationName].output
    ) {
      // no output defined = one-way operation so return empty response
      handled = true;
      callback("");
    }

    var result = operation(args, handleResult, options.headers, req);
    if (typeof result !== "undefined") {
      handleResult(null, result);
    }
  }

  _addWSSecurityHeader(headerElement) {
    var secElement = headerElement
      .element("wsse:Security")
      .attribute("soap:mustUnderstand", "1");

    secElement
      .attribute(
        "xmlns:wsse",
        "http://docs.oasis-open.org/wss/2004/01/" +
          "oasis-200401-wss-wssecurity-secext-1.0.xsd"
      )
      .attribute(
        "xmlns:wsu",
        "http://docs.oasis-open.org/wss/2004/01/" +
          "oasis-200401-wss-wssecurity-utility-1.0.xsd"
      );

    var now = new Date();
    var created = toXMLDate(now);
    var timeStampXml = "";

    var expires = toXMLDate(new Date(now.getTime() + 1000 * 600));

    var tsElement = secElement
      .element("wsu:Timestamp")
      .attribute("wsu:Id", "Timestamp-" + created);
    tsElement.element("wsu:Created", created);
    tsElement.element("wsu:Expires", expires);
  }

  _envelope(env, includeTimestamp) {
    env = env || XMLHandler.createSOAPEnvelope();

    if (includeTimestamp) {
      this._addWSSecurityHeader(env.header);
    }

    var soapHeaderElement = env.header;
    //add soapHeaders to envelope. Header can be xml, or JSON object which may or may not be described in WSDL/XSD.
    this.addSoapHeadersToEnvelope(soapHeaderElement, this.xmlHandler);
    return env;
  }

  _sendError(operation, error, callback, includeTimestamp) {
    var self = this,
      fault;

    var statusCode;
    if (error.Fault.statusCode) {
      statusCode = error.Fault.statusCode;
      error.Fault.statusCode = undefined;
    }

    var operationDescriptor = operation.describe(this.wsdl.definitions);
    debugDetail(
      "Server sendError. operationDescriptor: %j ",
      operationDescriptor
    );

    //get envelope descriptor
    var faultEnvDescriptor = operation.descriptor.faultEnvelope.elements[0];

    var soapNsURI = "http://schemas.xmlsoap.org/soap/envelope/";
    var soapNsPrefix = self.wsdl.options.envelopeKey || "soap";

    if (operation.soapVersion === "1.2") {
      soapNsURI = "http://www.w3.org/2003/05/soap-envelope";
    }

    var nsContext = self.createNamespaceContext(soapNsPrefix, soapNsURI);
    var envelope = XMLHandler.createSOAPEnvelope(soapNsPrefix, soapNsURI);

    //find the envelope body descriptor
    var bodyDescriptor = faultEnvDescriptor.elements[1];

    //there will be only one <Fault> element descriptor under <Body>
    var faultDescriptor = bodyDescriptor.elements[0];
    debugDetail("Server sendError. faultDescriptor: %j ", faultDescriptor);

    debug("Server sendError.  error.Fault: %j ", error.Fault);

    //serialize Fault object into XML as per faultDescriptor
    this.xmlHandler.jsonToXml(
      envelope.body,
      nsContext,
      faultDescriptor,
      error.Fault
    );

    self._envelope(envelope, includeTimestamp);
    var message = envelope.body.toString({ pretty: true });
    var xml = envelope.doc.end({ pretty: true });

    debug("Server sendError. Response envelope: %s ", xml);
    callback(xml, statusCode);
  }
}

module.exports = Server;
